Issue #110: update book tutorial (#112)
dotkernel/api-documentation: docs/book/v6/tutorials/create-book-module.md

Reviewer: Also, don't forget the "one sentence per line" rule!

```markdown
.
└── src/
    ├── Book/
    │   └── src/
    │       ├── Collection/
    │       │   └── BookCollection.php
    │       ├── Entity/
    │       │   └── Book.php
    │       ├── Handler/
    │       │   ├── GetBookCollectionHandler.php
    │       │   ├── GetBookHandler.php
    │       │   └── PostBookHandler.php
    │       ├── InputFilter/
    │       │   ├── Input/
    │       │   │   ├── AuthorInput.php
    │       │   │   ├── NameInput.php
    │       │   │   └── ReleaseDateInput.php
    │       │   └── CreateBookInputFilter.php
    │       ├── Service/
    │       │   ├── BookService.php
    │       │   └── BookServiceInterface.php
    │       ├── ConfigProvider.php
    │       └── RoutesDelegator.php
    └── Core/
        └── src/
            └── Book/
                └── src/
                    ├──Entity/
                    │   └──Book.php
                    ├──Repository/
                    │   └──BookRepository.php
                    └── ConfigProvider.php
    
```

> * `src/Book/src/Collection/BookCollection.php` - a collection refers to a container for a group of related objects, typically used to manage sets of related entities fetched from a database

Reviewer: Also, let's sort these items.

```markdown
.
└── src/
    ├── Book/
    │   └── src/
    │       ├── Collection/
    │       │   └── BookCollection.php

>     │       ├── Entity/

Reviewer: Remove node - the `Book` entity is in the `Core` module.

    │       │   └── Book.php
    │       ├── Handler/
    │       │   ├── GetBookCollectionHandler.php
    │       │   ├── GetBookResourceHandler.php
    │       │   └── PostBookResourceHandler.php
    │       ├── InputFilter/
    │       │   ├── Input/
    │       │   │   ├── AuthorInput.php
    │       │   │   ├── NameInput.php
    │       │   │   └── ReleaseDateInput.php
    │       │   └── CreateBookInputFilter.php
    │       ├── Service/
    │       │   ├── BookService.php
    │       │   └── BookServiceInterface.php
    │       ├── ConfigProvider.php
    │       └── RoutesDelegator.php
    └── Core/
        └── src/
            └── Book/
                └── src/
                    ├──Entity/
                    │   └──Book.php
                    ├──Repository/
                    │   └──BookRepository.php
                    └── ConfigProvider.php
    
```

> * `src/Book/src/Collection/BookCollection.php` – a collection refers to a container for a group of related objects, typically used to manage sets of related entities fetched from a database

Reviewer: By

> Also, let's sort these items.

and

> ... because that's how they appear in the IDE

I meant: sort these items both alphabetically AND hierarchically (maybe I should have made this clearer).
That meand first, by directory alphbetically and inside each directory the files are also sorted alphabetically - like you did in the above tree.

* `src/Book/src/ConfigProvider.php` – is a class that provides configuration for various aspects of the framework or application
* `src/Book/src/Handler/GetBookCollectionHandler.php` – handler that reflects the GET action for the BookCollection class
* `src/Book/src/Handler/GetBookResourceHandler.php` – handler that reflects the GET action for the Book entity
* `src/Book/src/Handler/PostBookResourceHandler.php` – handler that reflects the POST action for the Book entity
* `src/Book/src/InputFilter/CreateBookInputFilter.php` – input filters and validators
* `src/Book/src/InputFilter/Input/*` – input filters and validator configurations
* `src/Book/src/RoutesDelegator.php` – a routes delegator is a delegator factory responsible for configuring routing middleware based on routing configuration provided by the application

> * `src/Book/src/Service/BookService.php` – is a class or component responsible for performing a specific task or providing functionality to other parts of the application

Reviewer: ```suggestion
* `src/Book/src/Service/BookService.php` – is a class or component responsible for performing a specific task or providing functionality to other parts of the application
* `src/Book/src/Service/BookServiceInterface.php` – interface that reflects the publicly available methods in `BookService`
```

* `src/Core/src/Book/src/ConfigProvider.php` – is a class that provides configuration for Doctrine ORM
* `src/Core/src/Book/src/Entity/Book.php` – an entity refers to a PHP class that represents a persistent object or data structure
* `src/Core/src/Book/src/Repository/BookRepository.php` – a repository is a class responsible for querying and retrieving entities from the database

> ## Creating and configuring the module

Reviewer: This section could be a bit confusing for devs.
You instruct them to create a `ConfigProvider` with classes which don't exist yet - better leave it for when all the classes have been created.
We should keep here only the instructions for creating the directory structure and move the creation of the two `ConfigProvider`s and the `RouteDelegator` to the end of the `File creation and contents` section.

Firstly we will need the book module, so we will implement and create the basics for a module to be registered and functional.

In `src` and `src/Core/src` folders we will create one `Book` folder and in those we will create the `src` folder. So the final structure will be like this: `src/Book/src` and `src/Core/src/Book/src`.

In `src/Book/src` we will create 2 PHP files: `RoutesDelegator.php` and `ConfigProvider.php`. These files contain the necessary configurations.

> * `src/Book/src/RoutesDelegator.php`

Reviewer: This section is duplicate - you are creating the `RoutesDelegator` towards the end of the tutorial.
Let's keep only that one!

```markdown
.
└── src/
    ├── Book/
    │   └── src/
    │       ├── Collection/
    │       │   └── BookCollection.php
    │       ├── Entity/
    │       │   └── Book.php
    │       ├── Handler/
    │       │   ├── GetBookCollectionHandler.php
    │       │   ├── GetBookHandler.php
    │       │   └── PostBookHandler.php
    │       ├── InputFilter/
    │       │   └── CreateBookInputFilter.php
    │       ├── Service/
    │       │   ├── BookService.php
    │       │   └── BookServiceInterface.php
    │       ├── ConfigProvider.php
    │       └── RoutesDelegator.php
    ├── Core/
    │   └── src/
    │       └── Book/
    │           └── src/
    │               ├──Entity/
    │               │   └──Book.php
    │               ├──Repository/
    │               │   └──BookRepository.php
    │               └── ConfigProvider.php

>     └── App/

Reviewer: Move `App` module above the `Book` module because that's how they appear in the IDE.

* `src/Core/src/Book/src/Entity/Book.php` - an entity refers to a PHP class that represents a persistent object or data structure
* `src/Book/src/Handler/GetBookCollectionHandler.php` - handler that reflects the GET action for the BookCollection class
* `src/Book/src/Handler/GetBookHandler.php` - handler that reflects the GET action for the Book entity
* `src/Book/src/Handler/PostBookHandler.php` - handler that reflects the POST action for the Book entity
* `src/Core/src/Book/src/Repository/BookRepository.php` - a repository is a class responsible for querying and retrieving entities from the database
* `src/Book/src/Service/BookService.php` - is a class or component responsible for performing a specific task or providing functionality to other parts of the application
* `src/Book/src/ConfigProvider.php` -  is a class that provides configuration for various aspects of the framework or application
* `src/Book/src/RoutesDelegator.php` - a routes delegator is a delegator factory responsible for configuring routing middleware based on routing configuration provided by the application
* `src/Book/src/InputFilter/CreateBookInputFilter.php` - input filters and validators

> * `src/Core/src/App/src/InputFilter/Input/*` - input filters and validator configurations

Reviewer: ```suggestion
* `src/Book/src/InputFilter/Input/*` - input fields used by the input filters
```
Generic inputs are located under [src/App/src/InputFilter/Input/](https://github.com/dotkernel/api/tree/6.0/src/App/src/InputFilter/Input)
and module-specific inputs under [src/`<Module>`/src/InputFilter/Input/](https://github.com/dotkernel/api/tree/6.0/src/Admin/src/InputFilter/Input),
but never in the Core directory.

```php
<?php

declare(strict_types=1);

namespace Core\Book;

use Core\Book\Repository\BookRepository;
use Doctrine\ORM\Mapping\Driver\AttributeDriver;
use Dot\DependencyInjection\Factory\AttributedRepositoryFactory;

class ConfigProvider
{
    public function __invoke(): array
    {
        return [
            'dependencies'     => $this->getDependencies(),
            'doctrine'         => $this->getDoctrineConfig(),
        ];
    }

    private function getDependencies(): array
    {
        return [
            'factories' => [
                BookRepository::class => AttributedRepositoryFactory::class,
            ],
        ];
    }

    private function getDoctrineConfig(): array
    {
        return [
            'driver' => [
                'orm_default'  => [
                    'drivers' => [
                        'Core\Book\Entity' => 'BookEntities',
                    ],
                ],
                'BookEntities' => [
                    'class' => AttributeDriver::class,
                    'cache' => 'array',
                    'paths' => [__DIR__ . '/Entity'],
                ],
            ],
        ];
    }
}
```

> ### Registering the module

Reviewer: This section should also be moved further down to the point where the `ConfigProvider` has already been created.

                GetBookHandler::class           => [HandlerDelegatorFactory::class],
                GetBookCollectionHandler::class => [HandlerDelegatorFactory::class],
            ],
            'factories'  => [
                PostBookHandler::class          => AttributedServiceFactory::class,
                GetBookHandler::class           => AttributedServiceFactory::class,
                GetBookCollectionHandler::class => AttributedServiceFactory::class,
                BookService::class              => AttributedServiceFactory::class,
            ],
            'aliases'    => [
                BookServiceInterface::class => BookService::class,
            ],
        ];
    }

    private function getHalConfig(): array
    {
        return [
            AppConfigProvider::getResource(Book::class, 'book::view-book'),
            AppConfigProvider::getCollection(BookCollection::class, 'book::list-books', 'books'),
        ];
    }
}
```

* `src/Core/src/Book/src/ConfigProvider.php`

In `src/Core/src/Book/src` we will create 1 PHP file: `ConfigProvider.php`. This file contains the necessary configuration for Doctrine ORM.

```php
<?php

declare(strict_types=1);

> namespace Api\Book;

Reviewer: ```suggestion
namespace Core\Book;
```
Because you're in `src/Core/src/Book/src` not `src/Book/src`.

Author: This one is not in the `Core` directory

use Api\App\ConfigProvider as AppConfigProvider;
use Api\App\Factory\HandlerDelegatorFactory;
use Api\Book\Collection\BookCollection;
use Api\Book\Handler\GetBookCollectionHandler;
use Api\Book\Handler\GetBookHandler;
use Api\Book\Handler\PostBookHandler;
use Api\Book\Service\BookService;
use Api\Book\Service\BookServiceInterface;
use Core\Book\Entity\Book;
use Dot\DependencyInjection\Factory\AttributedServiceFactory;
use Mezzio\Application;
use Mezzio\Hal\Metadata\MetadataMap;

> class ConfigProvider

Reviewer: This ConfigProvider file needs to be redone from scratch because it references items from a non-Core ConfigProvider.

{
    public function __invoke(): array
    {
        return [
            'dependencies'     => $this->getDependencies(),
            MetadataMap::class => $this->getHalConfig(),
        ];
    }

>     private function getDependencies(): array

Reviewer: The Core ConfigProvider have nothing to do with handlers, in most cases they only register repositories.
See [src/Core/src/Admin/src/ConfigProvider.php](https://github.com/dotkernel/api/blob/6.0/src/Core/src/Admin/src/ConfigProvider.php) or [src/Core/src/User/src/ConfigProvider.php](https://github.com/dotkernel/api/blob/6.0/src/Core/src/User/src/ConfigProvider.php)

    public function getArrayCopy(): array
    {
        return [
            'uuid'        => $this->getUuid()->toString(),
            'name'        => $this->getName(),
            'author'      => $this->getAuthor(),
            'releaseDate' => $this->getReleaseDate(),
        ];
    }
}

```

* `src/Core/src/Book/src/Repository/BookRepository.php`

```php
<?php

declare(strict_types=1);

namespace Core\Book\Repository;

use Core\App\Repository\AbstractRepository;
use Core\Book\Entity\Book;
use Doctrine\ORM\QueryBuilder;
use Dot\DependencyInjection\Attribute\Entity;

#[Entity(name: Book::class)]
class BookRepository extends AbstractRepository
{

>     public function getBooks(array $params): QueryBuilder

Reviewer: ```suggestion
    public function getBooks(array $params, array $filters = []): QueryBuilder
```

namespace Api\Book\Service;

use Core\Book\Entity\Book;
use Core\Book\Repository\BookRepository;
use Doctrine\ORM\QueryBuilder;

interface BookServiceInterface
{
    public function getBookRepository(): BookRepository;

    public function saveBook(array $data): Book;

>     public function getBooks(array $filters = []): QueryBuilder;

Reviewer: ```suggestion
    public function getBooks(array $params): QueryBuilder;
```

namespace Api\Book\Service;

use Core\Book\Entity\Book;
use Core\Book\Repository\BookRepository;
use Doctrine\ORM\QueryBuilder;

interface BookServiceInterface
{
    public function getBookRepository(): BookRepository;

    public function saveBook(array $data): Book;

>     public function getBooks(array $params): QueryBuilder;

Reviewer: ```suggestion
    public function getBooks(array $params = []): QueryBuilder;
```

namespace Api\Book\Service;

use Core\App\Helper\Paginator;
use Core\Book\Entity\Book;
use Core\Book\Repository\BookRepository;
use DateTimeImmutable;
use Doctrine\ORM\QueryBuilder;
use Dot\DependencyInjection\Attribute\Inject;
use Exception;

class BookService implements BookServiceInterface
{
    #[Inject(BookRepository::class)]
    public function __construct(
        protected BookRepository $bookRepository
    ) {
    }

    public function getBookRepository(): BookRepository
    {
        return $this->bookRepository;
    }

    /**
     * @throws Exception
     */
    public function saveBook(array $data): Book
    {
        $book = new Book(
            $data['name'],
            $data['author'],
            new DateTimeImmutable($data['releaseDate'])
        );

        $this->bookRepository->saveResource($book);

        return $book;
    }

>     public function getBooks(array $filters = []): QueryBuilder

Reviewer: ```suggestion
    public function getBooks(array $params): QueryBuilder
```

    {

>         $params = Paginator::getParams($filters, 'book.created');

Reviewer: ```suggestion
        $filters = $params['filters'] ?? [];
        $params  = Paginator::getParams($params, 'book.created');

        $sortableColumns = [
            'book.name',
            'book.author',
            'book.releaseDate',
            'book.created',
        ];
        if (! in_array($params['sort'], $sortableColumns, true)) {
            $params['sort'] = 'book.created';
        }
```

use Doctrine\ORM\QueryBuilder;
use Dot\DependencyInjection\Attribute\Inject;
use Exception;

use function in_array;

class BookService implements BookServiceInterface
{
    #[Inject(BookRepository::class)]
    public function __construct(
        protected BookRepository $bookRepository
    ) {
    }

    public function getBookRepository(): BookRepository
    {
        return $this->bookRepository;
    }

    /**
     * @throws Exception
     */
    public function saveBook(array $data): Book
    {
        $book = new Book(
            $data['name'],
            $data['author'],
            new DateTimeImmutable($data['releaseDate'])
        );

        $this->bookRepository->saveResource($book);

        return $book;
    }

    public function getBooks(array $params = []): QueryBuilder
    {
        $filters = $params['filters'] ?? [];
        $params  = Paginator::getParams($filters, 'book.created');

        $sortableColumns = [
            'book.name',
            'book.author',
            'book.releaseDate',
            'book.created',
        ];

        if (! in_array($params['sort'], $sortableColumns, true)) {
            $params['sort'] = 'book.created';
        }

>         return $this->bookRepository->getBooks($params);

Reviewer: ```suggestion
        return $this->bookRepository->getBooks($params, $filters);
```

namespace Api\Book\Service;

use Core\App\Helper\Paginator;
use Core\Book\Entity\Book;
use Core\Book\Repository\BookRepository;
use DateTimeImmutable;
use Dot\DependencyInjection\Attribute\Inject;
use Exception;

> class BookService implements BookServiceInterface

Reviewer: This needs some refactoring, see [AdminService](https://github.com/dotkernel/api/blob/6.0/src/Admin/src/Service/AdminService.php) in `6.0` for example.
Then adapt `BookServiceInterface`.

}
```

* `src/Core/src/Book/src/Entity/Book.php`

> To keep things simple in this tutorial, our book will have 3 properties: `name`, `author` and `release date`.

Reviewer: ```suggestion
To keep things simple in this tutorial, our book will have 3 properties: `name`, `author` and `releaseDate`.
```

}
```

> We split all the inputs just for the purpose of this tutorial and to demonstrate a clean `CreateBookInputFilter` but you could have all the inputs created directly in the `CreateBookInputFilter` like this:

Reviewer: ```suggestion
We create separate `Input` files to demonstrate their reusability and obtain a clean `CreateBookInputFilter` but you could have all the inputs created directly in the `CreateBookInputFilter` like this:
```

    }
}
```

In `src/Book/src` we now create the 2 PHP files: `RoutesDelegator.php` and `ConfigProvider.php`.

`RoutesDelegator.php` contains all of our routes while `ConfigProvider` contains all the necessary configuration needed, so the above files work properly like dependency injection, aliases and so on.

* `src/Book/src/ConfigProvider.php`

```php
<?php

declare(strict_types=1);

namespace Api\Book;

use Api\App\ConfigProvider as AppConfigProvider;
use Api\App\Factory\HandlerDelegatorFactory;
use Api\Book\Collection\BookCollection;
use Api\Book\Handler\GetBookCollectionHandler;
use Api\Book\Handler\GetBookResourceHandler;
use Api\Book\Handler\PostBookResourceHandler;
use Api\Book\Service\BookService;
use Api\Book\Service\BookServiceInterface;
use Core\Book\Entity\Book;
use Dot\DependencyInjection\Factory\AttributedServiceFactory;
use Mezzio\Application;
use Mezzio\Hal\Metadata\MetadataMap;

class ConfigProvider
{
    public function __invoke(): array
    {
        return [
            'dependencies'     => $this->getDependencies(),
            MetadataMap::class => $this->getHalConfig(),
        ];
    }

>     private function getDependencies(): array

Reviewer: Format misaligned array values.

    {
        return [
            'factories' => [
                BookRepository::class => AttributedRepositoryFactory::class,
            ],
        ];
    }

    private function getDoctrineConfig(): array
    {
        return [
            'driver' => [
                'orm_default'  => [
                    'drivers' => [
                        'Core\Book\Entity' => 'BookEntities',
                    ],
                ],
                'BookEntities' => [
                    'class' => AttributeDriver::class,
                    'cache' => 'array',
                    'paths' => [__DIR__ . '/Entity'],
                ],
            ],
        ];
    }
}
```

### Registering the module

* register the module config by adding `Api\Book\ConfigProvider::class` and `Core\Book\ConfigProvider::class` in `config/config.php` under the `Api\User\ConfigProvider::class`
* register the namespace by adding this line `"Api\\Book\\": "src/Book/src/"` and `"Core\\Book\\": "src/Core/src/Book/src/"`, in composer.json under the autoload.psr-4 key
* update Composer autoloader by running the command:

```shell
composer dump-autoload
```

That's it. The module is now registered.

> We need to configure access to the newly created endpoints, add `books::list-books`, `book::view-book` and `book::create-book` to the authorization rbac array, under the `UserRole::ROLE_GUEST` key.

Reviewer: ```suggestion
We need to configure access to the newly created endpoints.
Open `config/autoload/authorization.global.php` and append the below route names to the `UserRoleEnum::Guest->value` key:

- `books::list-books`
- `book::view-book`
- `book::create-book`
```

> Make sure you read and understand the rbac [documentation](https://docs.dotkernel.org/dot-rbac-guard/v4/configuration/).

## Migrations

> We created the `Book` entity, but we didn't create the associated table for it.

Reviewer: In this section:

- fix typo `shel`
- replace `php bin/doctrine orm:validate-schema` with `php ./bin/doctrine orm:validate-schema`

```php
<?php

declare(strict_types=1);

namespace Core\Book;

use Core\Book\Repository\BookRepository;
use Doctrine\ORM\Mapping\Driver\AttributeDriver;
use Dot\DependencyInjection\Factory\AttributedRepositoryFactory;

class ConfigProvider
{
    public function __invoke(): array
    {
        return [
            'dependencies'     => $this->getDependencies(),
            'doctrine'         => $this->getDoctrineConfig(),
        ];
    }

    private function getDependencies(): array
    {
        return [
            'factories' => [
                BookRepository::class => AttributedRepositoryFactory::class,
            ],
        ];
    }

    private function getDoctrineConfig(): array
    {
        return [
            'driver' => [
                'orm_default'  => [
                    'drivers' => [
                        'Core\Book\Entity' => 'BookEntities',
                    ],
                ],
                'BookEntities' => [
                    'class' => AttributeDriver::class,
                    'cache' => 'array',
                    'paths' => [__DIR__ . '/Entity'],
                ],
            ],
        ];
    }
}
```

### Registering the module

> * register the module config by adding `Api\Book\ConfigProvider::class` and `Core\Book\ConfigProvider::class` in `config/config.php` under the `Api\User\ConfigProvider::class`

Reviewer: ```suggestion
* register the module config by adding `Api\Book\ConfigProvider::class,` and `Core\Book\ConfigProvider::class,` in `config/config.php` under the `Api\User\ConfigProvider::class,`
```

> * register the namespace by adding this line `"Api\\Book\\": "src/Book/src/"` and `"Core\\Book\\": "src/Core/src/Book/src/"`, in composer.json under the autoload.psr-4 key

Reviewer: ```suggestion
* register the namespace by adding this line `"Api\\Book\\": "src/Book/src/"` and `"Core\\Book\\": "src/Core/src/Book/src/"`, in `composer.json` under the `autoload`.`psr-4` key
```

* update Composer autoloader by running the command:

```shell
composer dump-autoload
```

That's it. The module is now registered.

We need to configure access to the newly created endpoints.
Open `config/autoload/authorization.global.php` and append the below route names to the `UserRoleEnum::Guest->value` key:

> * `books::list-books`

Reviewer: ```suggestion
* `book::list-books`
```

* `book::view-book`
* `book::create-book`

> Make sure you read and understand the rbac [documentation](https://docs.dotkernel.org/dot-rbac-guard/v4/configuration/).

## Migrations

We created the `Book` entity, but we didn't create the associated table for it.

> You can check the mapping files by running:

```shell
php ./bin/doctrine orm:validate-schema
```

Doctrine can handle the table creation, run the following command:

```shell
php ./vendor/bin/doctrine-migrations diff
```

This will check for differences between your entities and database structure and create migration files if necessary, in `src/Core/src/App/src/Migration`.

To execute the migrations run:

```shell
php ./vendor/bin/doctrine-migrations migrate
```

## Checking endpoints

> If we did everything as planned, we can call the `http://0.0.0.0:8080/book` endpoint and create a new book:

Reviewer: ```suggestion
First, we start a local server by executing:

composer serve

If we did everything as planned, we should be able to create a new book by executing the below command:
```

Wrap the `composer serve` command into markdown type hint of type `shell`.

        $this->getValidatorChain()
            ->attachByName(Date::class, [

>                 'message' => Message::INVALID_VALUE,

Reviewer: ```suggestion
                'message' => Message::invalidValue('releaseDate'),
```

use Core\App\Message;
use Dot\DependencyInjection\Attribute\Inject;
use Psr\Http\Message\ResponseInterface;
use Psr\Http\Message\ServerRequestInterface;
use Psr\Http\Server\RequestHandlerInterface;

class PostBookResourceHandler extends AbstractHandler implements RequestHandlerInterface
{
    #[Inject(
        CreateBookInputFilter::class,
        BookServiceInterface::class,
    )]
    public function __construct(
        protected CreateBookInputFilter $inputFilter,
        protected BookServiceInterface $bookService,
    ) {
    }

    public function handle(ServerRequestInterface $request): ResponseInterface
    {
        $this->inputFilter->setData((array) $request->getParsedBody());
        if (! $this->inputFilter->isValid()) {
            throw BadRequestException::create(
                detail: Message::VALIDATOR_INVALID_DATA,
                additional: ['errors' => $this->inputFilter->getMessages()]
            );
        }

        /** @var non-empty-array<non-empty-string, mixed> $data */
        $data = (array) $this->inputFilter->getValues();

        return $this->createdResponse($request, $this->bookService->createBook($data));
    }
}
```

> After we have the handler, we need to register some routes in the `RoutesDelegator` using our new grouping method, the same we created when we registered the module.

Reviewer: ```suggestion
After we have the handler, we need to register some routes in the `RoutesDelegator`, the same we created when we registered the module.
```
We abandoned route groups for this tutorial.

namespace Api\Book;

use Api\Book\Handler\GetBookCollectionHandler;
use Api\Book\Handler\GetBookHandler;
use Api\Book\Handler\PostBookHandler;
use Core\App\ConfigProvider;
use Dot\Router\RouteCollectorInterface;
use Mezzio\Application;
use Psr\Container\ContainerExceptionInterface;
use Psr\Container\ContainerInterface;
use Psr\Container\NotFoundExceptionInterface;

class RoutesDelegator
{
    /**
     * @throws ContainerExceptionInterface
     * @throws NotFoundExceptionInterface
     */

>     public function __invoke(ContainerInterface $container, string $serviceName, callable $callback): Application

Reviewer: As long as it's not justified to use groups, we should keep things simple because this way it's easier for devs to follow the tutorial.

Replace:

```php
$routeCollector->group('/book')->post('', PostBookHandler::class, 'book::create-book');
$routeCollector->group('/book/' . $uuid)->get('', GetBookHandler::class, 'book::view-book');
$routeCollector->group('/books')->get('', GetBookCollectionHandler::class, 'book::list-books');
```

with:

```php
$routeCollector->post('/book', PostBookHandler::class, 'book::create-book');
$routeCollector->get('/book/' . $uuid, GetBookHandler::class, 'book::view-book');
$routeCollector->get('/books', GetBookCollectionHandler::class, 'book::list-books');
```

    {
        $uuid = ConfigProvider::REGEXP_UUID;

        /** @var RouteCollectorInterface $routeCollector */
        $routeCollector = $container->get(RouteCollectorInterface::class);

        $routeCollector->group('/book')
            ->post('', PostBookHandler::class, 'book::create-book');

        $routeCollector->group('/book/' . $uuid)
            ->get('', GetBookHandler::class, 'book::view-book');

        $routeCollector->group('/books')
            ->get('', GetBookCollectionHandler::class, 'book::list-books');

        return $callback();
    }
}
```

We need to configure access to the newly created endpoints, add `books::list-books`, `book::view-book` and `book::create-book` to the authorization rbac array, under the `UserRole::ROLE_GUEST` key.

> > Make sure you read and understand the rbac documentation.

Reviewer: Add link to the documentation mentioned in this sentence so devs can find it easily:
https://docs.dotkernel.org/dot-rbac-guard/v4/configuration/

curl http://0.0.0.0:8080/books
```

> To retrieve a book, `curl` one of the links found in the output of the **list books** command, under `_embedded` . `books` . * . `_links` . `self` . `href`

Reviewer: ```suggestion
To fetch a book, `curl` one of the links found in the output of the **list books** command, under `_embedded` . `books` . * . `_links` . `self` . `href`.
The link should have the following format:
```
I replaced `retrieve` with `fetch`, because the latter is the more common term used when obtaining data via an API.
Also, added the last sentence so that the below curl snippet makes more sense.
